### ISSUE TITLE
`consult-buffer` slowness when using `consult-org-roam`

### ISSUE THREAD

**Issue by aavanian** [jgru/consult-org-roam#20]
I've been experimenting with `consult-org-roam` but find that `consult-buffer` becomes very slow once it's activated.

<details><summary>profiler result w/o `consult-org-roam`</summary>
<p>

```
         152  41% + #<compiled 0xe42e24cbeb30d77>
          45  12% + file-truename
          33   8% + #<compiled 0xe0ec30bfd1027e9>
          29   7% + #<compiled -0x1dff16fac04ec833>
          24   6% + read-from-minibuffer
          22   5%   Automatic GC
          10   2% + #<compiled 0x12cd0dc741558ae2>
          10   2% + file-relative-name
           7   1% + which-key--propertize-description
           5   1% + mapcar
           4   1% + marginalia--function-doc
           2   0% + vertico--update
           2   0% + #<compiled -0x1cf2072e101497a4>
           2   0% + all-completions
           1   0% + #<compiled 0x1dc193f24a4f>
           1   0% + which-key--key-description<
           1   0% + which-key--extract-key
           1   0% + advice--make-docstring
           1   0% + which-key--maybe-add-docstring
           1   0% + which-key--list-to-pages
           1   0% + which-key--echo
           1   0% + delete-window
           1   0% + consult--buffer-sort-visibility
           1   0% + abbreviate-file-name
           1   0% + powerline-render
           1   0% + #<compiled -0xf611ae1d4c5bae2>
           1   0% + timerp
           1   0% + #<compiled 0xf22d8c2e0fb90a9>
           1   0% + complete-with-action
           1   0% + seq-contains-p
           1   0% + mapc
           1   0% + vertico-sort-history-length-alpha
           1   0% + jit-lock-context-fontify
           1   0% + marginalia-annotate-binding
           1   0% + profiler-stop
```

</p>
</details>
<details><summary>profiler result w `consult-org-roam`</summary>
<p>

```
        8675  48% + consult-org-roam--remove-capture-dups
        5386  30% + file-truename
        1939  10% + file-relative-name
        1200   6%   Automatic GC
         227   1% + #<compiled -0x1dff16ed9485c833>
         147   0% + #<compiled 0xe42e24cbeb30d77>
          95   0% + #<compiled 0x1b94a7bbc16b79ff>
          64   0% + #<compiled 0xe0ec30bfd1027e9>
          45   0% + consult-org-roam--buffer-list-without-dups
          28   0% + read-from-minibuffer
          13   0% + #<compiled 0x12cd0dc741558ae2>
          11   0% + #<compiled -0x1db98559d8fa03f>
           7   0% + org-roam-buffer-list
           6   0% + all-completions
           5   0% + mapcar
           4   0% + org-get-limited-outline-regexp
           4   0% + apply
           4   0% + #<compiled 0xf22d8c2e0fb90a9>
           3   0% + replace-regexp-in-string
           3   0% + complete-with-action
           3   0% + #<compiled -0x1dff16ffaea7c833>
           2   0% + org-roam-descendant-of-p
           2   0% + consult--buffer-query
           2   0% + org-element--current-element
           2   0% + org-roam-file-p
           2   0% + org-element-keyword-parser
           2   0% + emacsql-escape-scalar
           2   0% + abbreviate-file-name
           2   0% + #<compiled -0x1cf2072e101497a4>
           2   0% + marginalia--function-doc
           2   0% + profiler-stop
           1   0% + default-font-width
           1   0% + org-roam-db--get-connection
           1   0% + outline-on-heading-p
           1   0% + org-element--collect-affiliated-keywords
           1   0% + assq-delete-all
           1   0% + smartparens-strict-mode
           1   0% + vertico--update
           1   0% + #<compiled -0x144835aa48fc5bcc>
           1   0% + sort
           1   0% + substitute-command-keys
           1   0% + timer-activate-when-idle
           1   0% + marginalia-annotate-binding
           1   0% + timer--time-less-p
           1   0% + timer-event-handler
```

</p>
</details>

In both cases, I start the profiler, call `consult-buffer` via `C-x b`, abort with `C-g` and stop the profiler.

I do have a lot of buffers (roam's: 265) open at the same time so looking into `consult-org-roam--buffer-list-without-dups`:
https://github.com/jgru/consult-org-roam/blob/b93cc856d1d2dd0719926e51c5a6309bbbf60b19/consult-org-roam-buffer.el#L95-L103

it seems to do a string match on `(length (org-roam-buffer-list))^2` (most of them are not duplicates).

It should probably faster to just look for `CAPTURE-.*` first then look for the original buffer?
Tough I don't fully understand the description of `consult-org-roam--buffer-list-without-dups` (seems a word is missing) so maybe something evades me.

**Comment by jgru** [comment]
Thank you for opening this issue, @aavanian. 

I corrected and clarified the comment in commit [4f5e77d](https://github.com/jgru/consult-org-roam/commit/4f5e77d36fb8256818df9042ee699954566d81d5). 
Your use case seems to be extreme, but the search for duplicates is indeed inefficient. 
Actually, there should only be very few duplicate buffers in the list returned by `org-roam-buffer-list`  so looking for buffers with prefix `CAPTURE-*` should be computationally more efficient. 
If you want to come up with an improved solution, it would be greatly appreciated. 

Best regards,
jgru

**Comment by jgru** [comment]
Hi @aavanian, 

I experimented with just removing the buffers that are prefixed with `CAPTURE-`.

Would you please test the following snippet whether it satisfies your performance requirements?

```elisp
(defun consult-org-roam--remove-capture-dups (buffer-list)
  "Remove CAPTURE-duplicates from BUFFER-LIST."
  (let ((out-list buffer-list))
  (dolist (x buffer-list)
    (when (s-starts-with-p "CAPTURE-" (buffer-name x))
      (setq out-list (delete x buffer-list))
      ))
    out-list))
```

Best regards,
jgru

**Comment by jgru** [comment]
Hi @aavanian, 

> I'm thinking the variable `org-roam-buffer-open-buffer-list` should be
> set when/during `consult-org-roam-buffer--get-roam-bufs` is called by
> the `consult` mechanics.

Indeed, thank you very much for this suggestion and drafting the code. 
I reviewed and tested your patch, which looks good. I merged it in commit 5d1e4a7. 

> 
> It's still slightly slower than I think it should but it's definitely
> usable for me now (I'll try to look further).
> 

I would be very happy if you would track this down even further and submit a PR (or a description of how to improve it even further). I rarely open more than 10 `org-roam``-buffers at once, so I did and do not experience any slowness.

Since you think that the current implementation is usable for you as well now, do you think that this issue can/should be closed?

> - shouldn't `(setq out-list (delete x buffer-list))` rather read
> `(setq out-list (delete x out-list))` (otherwise you lose the
> previous deletes)?

Definitely, I drafted this in a hurry, now it is corrected. 
  
> - shouldn't disabling `consult-org-roam-mode` remove
>   `org-roam-buffer-source` from `consult-buffer-sources`?

It should be removed, I think, and consider this to be a bug. 
I track this as #21. Thank you so much for pointing this out. 

Best regards,
jgru

**Comment by jgru** [comment]
Thank you for closing the issue, @aavanian. 
I just want to let you know that I added the removal of the `org-roam-buffer-source` from consult's `consult-buffer-sources` when disabling the minor mode in commit 96c95e5a14378cc6fc9c22981f5bfd9e18c419f6. #21 is closed now. 
I hope you enjoy using `consult-org-roam`.

Best regards,
jgru
